README.md
# scheduleit

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2019 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-08-20
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-08-20
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/scheduleit/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/scheduleit/compare/0.1.0...0.1.1


test/scheduleit/core_test.clj
(ns scheduleit.core-test
  (:require [clojure.test :refer :all]
            [scheduleit.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


utils.py
import itertools
import random
import collections
import operator
from functools import *
from itertools import *

def first(t):  return t[0]
def second(t): return t[1]
def third(t): return t[2]

def groupby(keyf,xs):
    res= {}
    for x in xs:
        k= keyf(x)
        if k in res:
            res[k].append(x)
        else:
            res[k] = [x]
    return res

def sortby (keyf,xs):
    xs.sort(key=keyf)
    return xs

def columns(tbl):
    width = len(tbl[0])
    height = len(tbl)
    return ([tbl[i][j] for i in range(height)] for j in range(width))

def rows (tbl):
    return (r for r in tbl)

def cells(tbl):
   width = len(tbl[0])
   height = len(tbl)
   return ([((i,j), tbl[i][j]) for i in range(height)] for j in range(width))

def colreduce(tbl,f,init):
    return reduce(f,columns(tbl),init)

def rowreduce(tbl,f,init):
    return reduce(f,rows(tbl),init)


#helper function for converting a sequence of [fieldnames, values]
#to a set of records.
def asRecords(xs):
    fields = xs[0]
    def makerec(vals):
        return {fields[i] : vals[i] for i in range(fields.count)}
    return map(makerec,xs[1:])

def srecord(name,src, compo, home, qty):
    return {"NAME"  : name,
            "SRC"   : src,
            "COMPO" : compo,
            "QTY"   : qty,
            "HOME"   : home}

def drecord(name, src, compo, qty, start, duration, dtype, overlap):
    return {"NAME"       : name,
            "SRC"        : src,
            "COMPO"      : compo,
            "QTY"        : qty,
            "START"      : start,
            "DURATION"   : duration,
            "DEMANDTYPE" : dtype,
            "OVERLAP"    : overlap,
            "DemandGroup" : dtype}

def distinctby(keyf,xs):
    return set(map(keyf,xs))

#wrapper around set, maybe useless.
def distinct(xs):
    return set(xs)
#transforms a list of pairs (l,r) into a list of (l r ..ln rn)
def unzip(pairs):
    res = []
    for l,r in pairs:
        res.append(l)
        res.append(r)
    return res

def rest(xs): return xs[1:]

def butlast(xs):
    return xs[:len(xs) - 1]

#from itertools recipes:
def pairwise(iterable):
    "s -> (s0,s1), (s1,s2), (s2, s3), ..."
    a, b = itertools.tee(iterable)
    next(b, None)
    return zip(a, b)

def nassoc(d,k,v):
    d[k] = v
    return d

def nassocP(d,kv):
    (k,v) = kv
    d[k] = v
    return d

def assoc(d,k,v, *kvs):
    res = d.copy()
    res[k] = v
    for k,v in pairwise(kvs):
        res[k] = v
    return res

def assocP(d,kv, *kvs):
    res = d.copy()
    k,v = kv
    res[k] = v
    for k,v in kvs:
        res[k] = v
    return res

def merge (*dicts):
    res = {}
    for d in dicts:
        reduce(lambda acc, kv: nassocP(acc,kv),d.items(),res)
    return res

def is_dict(x):
    return isinstance(x,dict)

##COW functions that return copies of collections.
#conjoin multiple items into a set.
def conj(s,x, *xs):
    if isdict(s):
        return assocP(s,x,xs)
    res = s.copy()
    res.add(s)
    reduce(lambda acc,x: res.add(x), xs, res)
    return res

#alias for itertools.chain
def concat(*collections):
    return chain(collections)

def mapcat(f,*collections):
    return concat(map(f,collections))

def unions(xs): return reduce(lambda l,r: l.union(r),xs)

def getKey(k):
    return lambda x : x[k]

def get(m,k):
    return m[k]
def get_keys(m,*ks):
    return map(lambda k: get(m,k),ks)

#itertools recipes.
def take(n, iterable):
    "Return first n items of the iterable as a list"
    return list(islice(iterable, n))

def tabulate(function, start=0):
    "Return function(0), function(1), ..."
    return map(function, count(start))


def consume(iterator, n):
    "Advance the iterator n-steps ahead. If n is none, consume entirely."
    # Use functions that consume iterators at C speed.
    if n is None:
        # feed the entire iterator into a zero-length deque
        collections.deque(iterator, maxlen=0)
    else:
        # advance to the empty slice starting at position n
        next(islice(iterator, n, n), None)

def nth(iterable, n, default=None):
    "Returns the nth item or a default value"
    return next(islice(iterable, n, None), default)

def quantify(iterable, pred=bool):
    "Count how many times the predicate is true"
    return sum(map(pred, iterable))

def padnone(iterable):
    """Returns the sequence elements and then returns None indefinitely.

    Useful for emulating the behavior of the built-in map() function.
    """
    return chain(iterable, repeat(None))

def ncycles(iterable, n):
    "Returns the sequence elements n times"
    return chain.from_iterable(repeat(tuple(iterable), n))

def dotproduct(vec1, vec2):
    return sum(map(operator.mul, vec1, vec2))

def flatten(listOfLists):
    "Flatten one level of nesting"
    return chain.from_iterable(listOfLists)

def repeatfunc(func, times=None, *args):
    """Repeat calls to func with specified arguments.

    Example:  repeatfunc(random.random)
    """
    if times is None:
        return starmap(func, repeat(args))
    return starmap(func, repeat(args, times))

def grouper(iterable, n, fillvalue=None):
    "Collect data into fixed-length chunks or blocks"
    # grouper('ABCDEFG', 3, 'x') --> ABC DEF Gxx"
    args = [iter(iterable)] * n
    return zip_longest(*args, fillvalue=fillvalue)

def roundrobin(*iterables):
    "roundrobin('ABC', 'D', 'EF') --> A D E B F C"
    # Recipe credited to George Sakkis
    pending = len(iterables)
    nexts = cycle(iter(it).__next__ for it in iterables)
    while pending:
        try:
            for next in nexts:
                yield next()
        except StopIteration:
            pending -= 1
            nexts = cycle(islice(nexts, pending))

def partition(pred, iterable):
    'Use a predicate to partition entries into false entries and true entries'
    # partition(is_odd, range(10)) --> 0 2 4 6 8   and  1 3 5 7 9
    t1, t2 = tee(iterable)
    return filterfalse(pred, t1), filter(pred, t2)

def powerset(iterable):
    "powerset([1,2,3]) --> () (1,) (2,) (3,) (1,2) (1,3) (2,3) (1,2,3)"
    s = list(iterable)
    return chain.from_iterable(combinations(s, r) for r in range(len(s)+1))

def unique_everseen(iterable, key=None):
    "List unique elements, preserving order. Remember all elements ever seen."
    # unique_everseen('AAAABBBCCDAABBB') --> A B C D
    # unique_everseen('ABBCcAD', str.lower) --> A B C D
    seen = set()
    seen_add = seen.add
    if key is None:
        for element in filterfalse(seen.__contains__, iterable):
            seen_add(element)
            yield element
    else:
        for element in iterable:
            k = key(element)
            if k not in seen:
                seen_add(k)
                yield element

# def unique_justseen(iterable, key=None):
#     "List unique elements, preserving order. Remember only the element just seen."
#     # unique_justseen('AAAABBBCCDAABBB') --> A B C D A B
#     # unique_justseen('ABBCcAD', str.lower) --> A B C A D
#     return map(next, map(itemgetter(1), groupby(iterable, key)))

def iter_except(func, exception, first=None):
    """ Call a function repeatedly until an exception is raised.

    Converts a call-until-exception interface to an iterator interface.
    Like builtins.iter(func, sentinel) but uses an exception instead
    of a sentinel to end the loop.

    Examples:
        iter_except(functools.partial(heappop, h), IndexError)   # priority queue iterator
        iter_except(d.popitem, KeyError)                         # non-blocking dict iterator
        iter_except(d.popleft, IndexError)                       # non-blocking deque iterator
        iter_except(q.get_nowait, Queue.Empty)                   # loop over a producer Queue
        iter_except(s.pop, KeyError)                             # non-blocking set iterator

    """
    try:
        if first is not None:
            yield first()            # For database APIs needing an initial cast to db.first()
        while 1:
            yield func()
    except exception:
        pass

def first_true(iterable, default=False, pred=None):
    """Returns the first true value in the iterable.

    If no true value is found, returns *default*

    If *pred* is not None, returns the first item
    for which pred(item) is true.

    """
    # first_true([a,b,c], x) --> a or b or c or x
    # first_true([a,b], x, f) --> a if f(a) else b if f(b) else x
    return next(filter(pred, iterable), default)

##def random_product(*args, repeat=1):
##    "Random selection from itertools.product(*args, **kwds)"
##    pools = [tuple(pool) for pool in args] * repeat
##    return tuple(random.choice(pool) for pool in pools)

def random_permutation(iterable, r=None):
    "Random selection from itertools.permutations(iterable, r)"
    pool = tuple(iterable)
    r = len(pool) if r is None else r
    return tuple(random.sample(pool, r))

def random_combination(iterable, r):
    "Random selection from itertools.combinations(iterable, r)"
    pool = tuple(iterable)
    n = len(pool)
    indices = sorted(random.sample(range(n), r))
    return tuple(pool[i] for i in indices)

def random_combination_with_replacement(iterable, r):
    "Random selection from itertools.combinations_with_replacement(iterable, r)"
    pool = tuple(iterable)
    n = len(pool)
    indices = sorted(random.randrange(n) for i in range(r))
    return tuple(pool[i] for i in indices)


monthly.py
#-------------------------------------------------------------------------------
# Name:        module1
# Purpose:
#
# Author:      tom
#
# Created:     23/05/2019
# Copyright:   (c) tom 2019
# Licence:     <your licence>
#-------------------------------------------------------------------------------

from itertools import product
import utils
import pickle
from pulp import *

def dictvals(d):
    return {k : value(d[k]) for k in d.keys()}

#rough packaging of information we'll use to
#define a MIP problem instance.  We'll refine this
#to closely match the data we pull out of Excel going forward.
def sampleData(small=True):

    ##This is just some lame stuff.
    #total number of MTT teams
    if small == True:
        MTT = 10
    else:
        MTT = 10
    #total weeks
    if small == True:
        WKS = 156
    else:
        WKS = 156
    years = WKS//52
    #mtt teams
    ms = range(MTT)
    #weeks
    ws = range(WKS)
    #unit types
    types = {'cst','dcrf','cre','sr','hrf','cerfp'}

    #mapping of unit->type
    ##this is one cough-syrup addled way to do it
    ##with dictionary comprehensions.
    csts   = {x:'cst' for x in range(0,55)}
    dcrfs  = {x:'cst' for x in range(55,61)}
    cres   = {x:'cre' for x in range(61, 65)}
    srs    = {x:'sr' for x in range(65,77)}
    hrfs   = {x:'hrf' for x in range(77,87)}
    cerfps = {x:'cerfp' for x in range(87,104)}

    #assuming we have a table of id->type
    unit_type = utils.merge(csts,dcrfs,cres,srs,hrfs,cerfps)

    #lame total unit count
    if small == True :
        total_units = 10
    else:
        total_units = len(unit_type)
    #range of unit ids...
    us = range(total_units)
    # training interval relative to type, weeks
    interval_type =  {'sr'   :13,
                      'cre'  :52,
                      'dcrf' :52,
                      'cst'  :78,
                      'hrf'  :78,
                      'cerfp':78}
    ys = range(years)

    constant_demands = {'sr':6, 'dcrf':6, 'hrf':10,'cerfp':17}

    static_demands = \
        {(t,y):constant_demands.get(t)
         for t,y in utils.product(types,ys)
         if t in constant_demands}

    variable_demands = {('cre',1):1,('cre',2):4,('cre',3):2}

    #all the yearly demands.
    #yearly demands can be trivially translated to weekly
    #demands...
    yearly_demands = utils.merge(variable_demands, static_demands)

    return {'MTT':MTT,
            'WKS':WKS,
            'ms':ms,
            'ws':ws,
            'types':types,
            'total_units':total_units,
            'us':range(total_units),
            'unit_type':unit_type,
            'interval_type':interval_type,
            'msn_demand':yearly_demands}

#munge our input into a dictionary we can use
def readInputData():
    return sampleData()

###this helps recover our solution in an ordered fashion.

#given a set of parameters, defines an instance of a scheduling model using
#the PuLP modelling functions.  As stated, PuLP will use its built in
#solver.
def buildModel(params):
#asumming we're passed a dictionary of requisite
#parameters, we can unpack it and construct a model
#from the inputs.

#bind the parameters locally for easier reference.
    (MTT,      #scalar, Total number of training teams
    ms,       #numerical range of [0..MTT)
    WKS,      #scalr, total number of weeks to schedule
    ws,       #numerical range of [0..WKS)
    types,    #set of unit types
    total_units, #scalar count of units
    us,         #numerical range [0..total_units)
    unit_type,  #unitid->type
    interval_type,   #unit_type->training-interval
    #all the yearly demands.
    #yearly demands can be trivially translated to weekly
    #demands...
    msn_demand) = utils.get_keys(params,'MTT','ms','WKS','ws','types',
    'total_units','us','unit_type', 'interval_type','msn_demand')

##    #setting up a minimization.
    sched = LpProblem("Sched", LpMinimize)

    months_weeks = 12/52

    #scale the weeks to months....
    ws = range(WKS * months_weeks)
    #scale the intervals to months....
    interval_type = {k:weeks * months_weeks for k,weeks in interval_type.items()}

    #variables

    print(len(ws))
## assign(m,u,w) # binary decision variable determining whether training
## team m is assigned to train unit u, during week w of the schedule.


   #This lets us say "we want to create family of variables, across the
   #index m_u_w, with similar properties (in this case they're binary)
    assign = LpVariable.dicts('assign',utils.product(ms,us,ws),
                              lowBound = 0.0, upBound = 1.0, cat=LpInteger
                              )

    #Objective function: z = assigncost = sum(m,u,w)assign(m,u,w)
    #min z, e.g. minimize the total number of assignments
    #For now, we reflect no higher order notion of cost within the parameters.
    #One could easily envision certain combinations being more desirable,
    #hence the inclusion of a weight parameter
    assigncost = LpVariable('assigncost')
    sched += assigncost == lpSum(assign[(m,u,w)]*1.0
        for (m,u,w) in utils.product(ms,us,ws))

    #helper variables:
    #mtt training events per week
    #events(m,w) = sum(u)assign(m,u,w) forall m in ms,w in ws, u in us
    events = LpVariable.dicts("events",utils.product(ms,ws))
    for (m,w) in events.keys():
        sched += events[m,w]  == lpSum((assign[m,u,w] for u in us))

    #only one training event allowed per week
    for (m,w) in events.keys():
        sched += events[m,w] <= 1

    #add a var to track unit training

    trained = LpVariable.dicts("trained",utils.product(us,ws))
    for (u,w) in trained.keys():
        sched += trained[u,w] == lpSum((assign[m,u,w] for m in ms))

    #unit training credits by interval
    #We account for training intervals by
    #recording wait times as a function of
    #max training intervals and previous
    #wait times.  We introduce the linear
    #decision variable, wait(u,w), which
    #indicates how long unit u has been waiting
    #for training at week w.

    #we may want to add in some minimum frequency too, to ensure
    #we don't allow things like month-month training, which is likely
    #infeasible in practice.  Something like, 1/2 the interval is the
    #lower bound.

    wait = LpVariable.dicts("wait",utils.product(us,ws),lowBound = 0.0)
    for u in us:
        interval = interval_type[unit_type[u]]
        for w in utils.butlast(ws):
            sched += wait[u,w + 1] == wait[u,w] + 1 - trained[u,w]*(interval + 1)

    #we can train a unit more frequently, but not lapse in training.
    #Wait times cannot exceed mandated training intervals.
    #note: we can add a deviation goal to this later if we want
    #to or need to relax this.  Alternately, include in objective...
    for (u,w) in wait.keys():
        sched += wait[u,w] <= interval_type[unit_type[u]]

    #goal is to minimize the total number of assignments we need.
    #This should induce maximal wait times as well.
    sched.setObjective(assigncost)
    def getsolution():
        res = {}
        for (m,u,w) in assign.keys():
            if value(assign[m,u,w]) > 0.0 :
                res[m,u,w] = value(assign[m,u,w])
        return {'assigned':res, #dictvals(assign),
                'trained':dictvals(trained),
                'wait':dictvals(wait),
                'assigncost':value(assigncost)}
    return (sched,getsolution)


###keep track of our last solve, since this can take some time.
lastsolution = None

###Reads external data for parameters necessary to define TMAS, builds the model,
###solves the model, prints results.  Note: since we have Python data structures
###here, it should be really easy to pipe the results into post processing and
###other functions, using all of the Python facilities for operating on data.
def main(run = False):
    global lastsolution
    #should be replaced with IO functions to actual data
    if run:
        params = sampleData(small=False)
        sched,getsolution = buildModel(params)
        print("Solving model....")
        sched.solve()
        print("Status:", LpStatus[sched.status])
        res = getsolution()
        lastsolution = res
        return res
##
##def spit(obj,path = "obj.py"):
##    with open(path, 'wb') as outfile:
##        pickle.dump(obj,outfile)
##
##def load(path= "obj.py"):
##    with open(path, "rb") as infile:
##        return pickle.load(infile)
##
##def writelp():
##    params = inputDataToParams(readInputData())
##    squirm,getsolution = buildModel(params)
##    print("Saving model as SqurimModel.lp....")
##    squirm.writeLP("SquirmModel.lp")
##
##
##def restable(res):
##    return post.asTable(post.getRecords(res))

schedule.py
#-------------------------------------------------------------------------------
# Name:        module1
# Purpose:
#
# Author:      tom
#
# Created:     23/05/2019
# Copyright:   (c) tom 2019
# Licence:     <your licence>
#-------------------------------------------------------------------------------

from itertools import product
import utils
import pickle
from pulp import *

def dictvals(d):
    return {k : value(d[k]) for k in d.keys()}

#rough packaging of information we'll use to
#define a MIP problem instance.  We'll refine this
#to closely match the data we pull out of Excel going forward.
def sampleData(small=True):

    ##This is just some lame stuff.
    #total number of MTT teams
    if small == True:
        MTT = 1
    else:
        MTT = 10
    #total weeks
    if small == True:
        WKS = 156
    else:
        WKS = 156
    years = WKS//52
    #mtt teams
    ms = range(MTT)
    #weeks
    ws = range(WKS)
    #unit types
    types = {'cst','dcrf','cre','sr','hrf','cerfp'}

    #mapping of unit->type
    ##this is one cough-syrup addled way to do it
    ##with dictionary comprehensions.
    csts   = {x:'cst' for x in range(0,55)}
    dcrfs  = {x:'dcrf' for x in range(55,61)}
    cres   = {x:'cre' for x in range(61, 65)}
    srs    = {x:'sr' for x in range(65,77)}
    hrfs   = {x:'hrf' for x in range(77,87)}
    cerfps = {x:'cerfp' for x in range(87,104)}

    #assuming we have a table of id->type
    unit_type = utils.merge(csts,dcrfs,cres,srs,hrfs,cerfps)

    #lame total unit count
    if small == True :
        total_units = 1
    else:
        total_units =  len(unit_type)
    #range of unit ids...
    us = range(total_units)
    # training interval relative to type, weeks
    interval_type =  {'sr'   :13,
                      'cre'  :52,
                      'dcrf' :52,
                      'cst'  :78,
                      'hrf'  :78,
                      'cerfp':78}
    ys = range(years)

    constant_demands = {'sr':6, 'dcrf':6, 'hrf':10,'cerfp':17}

    static_demands = \
        {(t,y):constant_demands.get(t)
         for t,y in utils.product(types,ys)
         if t in constant_demands}

    variable_demands = {('cre',1):1,('cre',2):4,('cre',3):2}

    #all the yearly demands.
    #yearly demands can be trivially translated to weekly
    #demands...
    yearly_demands = utils.merge(variable_demands, static_demands)

    return {'MTT':MTT,
            'WKS':WKS,
            'ms':ms,
            'ws':ws,
            'types':types,
            'total_units':total_units,
            'us':range(total_units),
            'unit_type':unit_type,
            'interval_type':interval_type,
            'msn_demand':yearly_demands}

#munge our input into a dictionary we can use
def readInputData():
    return sampleData()

###this helps recover our solution in an ordered fashion.

#given a set of parameters, defines an instance of a scheduling model using
#the PuLP modelling functions.  As stated, PuLP will use its built in
#solver.
def buildModel(params):
#asumming we're passed a dictionary of requisite
#parameters, we can unpack it and construct a model
#from the inputs.

#bind the parameters locally for easier reference.
    (MTT,      #scalar, Total number of training teams
    ms,       #numerical range of [0..MTT)
    WKS,      #scalr, total number of weeks to schedule
    ws,       #numerical range of [0..WKS)
    types,    #set of unit types
    total_units, #scalar count of units
    us,         #numerical range [0..total_units)
    unit_type,  #unitid->type
    interval_type,   #unit_type->training-interval
    #all the yearly demands.
    #yearly demands can be trivially translated to weekly
    #demands...
    msn_demand) = utils.get_keys(params,'MTT','ms','WKS','ws','types',
    'total_units','us','unit_type', 'interval_type','msn_demand')

##    #setting up a minimization.
    sched = LpProblem("Sched", LpMinimize)
    #variables


    #add a var to track unit training

    ##introduce a binary decision variable to decide whether
    ##team m trained someone during week w:
    trained = LpVariable.dicts('trained',utils.product(ms,ws),
                                lowBound = 0.0, upBound = 1, cat=LpInteger)

    #for w in ws:
    #    sched += lpSum(trained[m,w] for m in ms) <= 1

    #We have a linear variable that only takes on binary values
    #of 0 or 1 to indicate the assignment of mtt m, to unit u,
    #at time w.  Since unit_assigned can only take on non-zero
    #values when trained(u,w) is 1 for the unit, this leads to
    #a singular value for unit_assigned which is identical
    #to the index of u in trained(u,w).  If the unit is not
    #trained, then unit_assigned will be 0, or unassigned.
    #So we can coerce our implicitly integral value for
    #unit_assigned by dividing the the related unit index constant.
    #Summing across all assigned indices (either 0 or a unit
    #index defined by train(u,w)), we then have 1 possibility
    #for non-zero assignment.  Dividing by the unit index
    #projects the index onto the set {0,1}, giving us an implicit
    #integer variable.
    assign = LpVariable.dicts('assign',utils.product(ms,us,ws),
                              lowBound = 0.0, upBound = 1)
    for (m,w) in product(ms,ws):
        #number of units assigned to m at w must by = whether m trained at w
        sched +=  lpSum(assign[m,u,w] for u in us) == trained[m,w]
        for u in us:
            #unit assigned to m at w must be either 0 or 1, depending
            #on trained.  If trained is 0, assign == 0, else
            #if trained is 1, assign is either 0|1.
            sched += assign[m,u,w] <= trained[m,w]


    #Objective function: z = assigncost = sum(m,u,w)assign(m,u,w)
    #min z, e.g. minimize the total number of assignments
    #For now, we reflect no higher order notion of cost within the parameters.
    #One could easily envision certain combinations being more desirable,
    #hence the inclusion of a weight parameter
    assigncost = LpVariable('assigncost')
    sched += assigncost == lpSum(assign[(m,u,w)]*1.0
        for (m,u,w) in assign.keys())

    #helper variables:
    #mtt training events per week
    #events(m,w) = sum(u)assign(m,u,w) forall m in ms,w in ws, u in us
#I think this is controlled directly by trained now...
    #events = LpVariable.dicts("events",utils.product(ms,ws))
    #for (m,w) in events.keys():
    #    sched += events[m,w]  == lpSum((assign[m,u,w] for u in us))

    #only one training event allowed per week
#    for (m,w) in events.keys():
#        sched += events[m,w] <= 1

    #unit training credits by interval
    #We account for training intervals by
    #recording wait times as a function of
    #max training intervals and previous
    #wait times.  We introduce the linear
    #decision variable, wait(u,w), which
    #indicates how long unit u has been waiting
    #for training at week w.

    #we may want to add in some minimum frequency too, to ensure
    #we don't allow things like month-month training, which is likely
    #infeasible in practice.  Something like, 1/2 the interval is the
    #lower bound.

    unit_trained = LpVariable.dicts("unit_trained",utils.product(us,ws),lowBound = 0.0, upBound = 1.0)
    for (u,w) in unit_trained.keys():
        sched += unit_trained[u,w] == lpSum(assign[m,u,w] for m in ms)

    wait = LpVariable.dicts("wait",utils.product(us,ws),lowBound = 0.0)
    for u in us:
        #our unit_type index is 0-based, we need to offset by 1 in the
        #formulation.
        interval = interval_type[unit_type[u]]
        for w in utils.butlast(ws):
            sched += wait[u,w + 1] == wait[u,w] + 1 - unit_trained[u,w]*(interval + 1)

    #we can train a unit more frequently, but not lapse in training.
    #Wait times cannot exceed mandated training intervals.
    #note: we can add a deviation goal to this later if we want
    #to or need to relax this.  Alternately, include in objective...

    for (u,w) in wait.keys():
        #we need to offset our unit index, since unit_type is 0 based.
        sched += wait[u,w] <= interval_type[unit_type[u]]

    waitcost = LpVariable('waitcost')
    sched += waitcost == lpSum(wait[u,w] for (u,w) in wait.keys())


    objective = LpVariable('assign plus wait')

    #we want to minimize assignments, maximize wait, so use negative
    #wait and emphasize the importance of assignment.

    sched += objective == 100 * assigncost - waitcost

    #goal is to minimize the total number of assignments we need.
    #This should induce maximal wait times as well.
    sched.setObjective(objective)
    #sched.setObjective(assigncost)
    def getsolution():
        res = {}
        for (m,u,w) in assign.keys():
            if value(assign[m,u,w]) > 0.0 :
                res[m,u,w] = value(assign[m,u,w])
        return {'assigned':res, #dictvals(assign),
                'trained':dictvals(trained),
                'wait':dictvals(wait),
                'assigncost':value(assigncost),
                'waitcost':value(waitcost),
                'objective':value(objective)}
    return (sched,getsolution)


###keep track of our last solve, since this can take some time.
lastsolution = None

###Reads external data for parameters necessary to define TMAS, builds the model,
###solves the model, prints results.  Note: since we have Python data structures
###here, it should be really easy to pipe the results into post processing and
###other functions, using all of the Python facilities for operating on data.
def main(run = False):
    global lastsolution
    #should be replaced with IO functions to actual data
    if run:
        params = sampleData(small=False)
        sched,getsolution = buildModel(params)
        print("Solving model....")
        sched.solve()
        print("Status:", LpStatus[sched.status])
        res = getsolution()
        lastsolution = res
        return res

def instance(small = False):
    params = sampleData(small=small)
    sched,getsolution = buildModel(params)
    return {"model":sched, "solution":getsolution}
##
##def spit(obj,path = "obj.py"):
##    with open(path, 'wb') as outfile:
##        pickle.dump(obj,outfile)
##
##def load(path= "obj.py"):
##    with open(path, "rb") as infile:
##        return pickle.load(infile)
##
##def writelp():
##    params = inputDataToParams(readInputData())
##    squirm,getsolution = buildModel(params)
##    print("Saving model as SqurimModel.lp....")
##    squirm.writeLP("SquirmModel.lp")
##
##
##def restable(res):
##    return post.asTable(post.getRecords(res))
